### PR TITLE
cellL10n: Implement SJIS<->Unicode conversions.

### PR DESCRIPTION
**Pull request by achurch**
Needed for Tales of Vesperia.

I tried to avoid breaking the Windows code, but I don't have a Windows environment to build in, so someone will need to check that that code still works.

### CODE REVIEW

File: rpcs3/Emu/SysCalls/Modules/cellL10n.cpp [hunk 1]
```diff
@@ -205,48 +203,63 @@ s32 _L10nConvertStr(s32 src_code, const void *src, size_t * src_len, s32 dst_cod
...
-	std::string wrapped_source((char*)src);
-	//if (wrapped_source.length != src_len.GetValue()) return SRCIllegal;
+#ifdef _MSC_VER
+	std::string wrapped_source = std::string(src, src_len);
 	std::string target = _OemToOem(srcCode, dstCode, wrapped_source);
 
 	if (target.length() > *dst_len) return DSTExhausted;
 
-	Memory.WriteString(dst.addr(), target);
+	memcpy(dst.get_ptr(), target.c_str(), target.length());
```
Comment (danilaml) [inline rpcs3/Emu/SysCalls/Modules/cellL10n.cpp]: `dst.get_ptr()` ? dst is `void *` here.

File: rpcs3/Emu/SysCalls/Modules/cellL10n.cpp [hunk 2]
```diff
@@ -205,48 +203,63 @@ s32 _L10nConvertStr(s32 src_code, const void *src, size_t * src_len, s32 dst_cod
 		|| ((!dst_page_converted) && (dstCode == 0)))
 		return ConverterUnknown;
 
-	if (strnlen_s((char*)src, *src_len) != *src_len) return SRCIllegal;
-	//std::string wrapped_source = (char*)Memory.VirtualToRealAddr(src.addr());
-	std::string wrapped_source((char*)src);
-	//if (wrapped_source.length != src_len.GetValue()) return SRCIllegal;
+#ifdef _MSC_VER
+	std::string wrapped_source = std::string(src, src_len);
```
Comment (danilaml) [inline rpcs3/Emu/SysCalls/Modules/cellL10n.cpp]: There is no string constructor for (const void*, s32)

Comment (achurch) [inline rpcs3/Emu/SysCalls/Modules/cellL10n.cpp]: Fixed.

File: rpcs3/Emu/SysCalls/Modules/cellL10n.cpp (hunk 1 above)
Comment (achurch) [inline rpcs3/Emu/SysCalls/Modules/cellL10n.cpp]: Fixed.

**Comment by tambry** [comment]
@achurch Could you please rebase?

**Comment by achurch** [comment]
Done.

**Comment by tambry** [comment]
@achurch That's a merge not a rebase.
See [git rebase](https://git-scm.com/docs/git-rebase).

**Comment by achurch** [comment]
Sorry, force of habit.

**Comment by tambry** [comment]
@achurch No problem. I'll take a look at if it works on Windows sometime tomorrow.

**Comment by tambry** [comment]
I seem to have no games that uses any SJIS <-> Unicode functions. Most of those games that use those functions seem to be Japanese.
@B1ackDaemon Maybe you could test and see if it works? (some likely games: Hyperdimension Neptunia, Tales of Vesperia, Trinity Universe, [others](https://github.com/Zangetsu38/Log-Rpcs3/search?p=3&q=SJISstoUTF8s&type=Code&utf8=%E2%9C%93))

**Comment by tambry** [comment]
@achurch
Needs a rebase.

I found that Minecraft [NPUB31419] uses l10n_get_converter and l10n_convert_str of the string conversion functions.
This PR makes Minecraft crash. Though previously it would simply make it crash later on in some DX12 string handling, which could've been caused by the problems of the previous string conversion.

**Log:**
https://gist.githubusercontent.com/tambry/1a12b5ab89082b181388/raw/4bc2a45c87a5e4b72918c86454e46c0d49388acf/Minecraft%2520%255BNPUB31419%255D

**Comment by achurch** [comment]
It looks like Minecraft is passing NULL as the destination pointer; it may be that the function treats dst==NULL as "return only the length and discard the data". I'll go ahead and add that in, since I don't see any other reason the program would be passing a null pointer.

**Comment by achurch** [comment]
FWIW, it does indeed look like Minecraft is using a "get length first" pattern in a previous call (which doesn't crash for whatever reason):
```
HLE: W {PPU Thread[0x1] (main_thread)[0x0081e1b4]} cellL10n: l10n_convert_str(cd=65536, src=*0x20df0580, src_len=*0xd000f9e0, dst=*0x0, dst_len=*0xd000f9e4)
HLE: W {PPU Thread[0x1] (main_thread)[0x0081e1b4]} cellL10n: l10n_convert_str(cd=65536, src=*0x20df0580, src_len=*0xd000f9e0, dst=*0x22ded150, dst_len=*0xd000f9e4)
```

Given that that previous call doesn't crash, it may be that the null pointer isn't the cause of the crash, but the call pattern does suggest to me that this special handling of a null destination pointer is correct.

**Comment by tambry** [comment]
@achurch
From the documentation for l10n_convert_str:
```
If NULL is specified to dst, the buffer size required to store the conversion result will be stored to dst_len.
```

**Comment by tambry** [comment]
@achurch Needs to rebased.

**Comment by achurch** [comment]
Done.

**Pull request merged**